### PR TITLE
[FEA] add more routines to partition graph

### PR DESCRIPTION
**Pull request by stadlmax**
# Modulus Pull Request

## Description
- extends the previous design of `GraphPartition` which already had more generic partitioning schemes in mind
- besides the previous "naive" partitioning scheme, this PR now adds support for defining a graph partition by providing
a mapping for each node ID to a rank
- based on this, this PR would also add support for using coordinates to define these mapping leveraging domain information to reduce communication between partitions

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/models/gnn_layers/distributed_graph.py [hunk 1]
```diff
@@ -117,94 +131,111 @@ def partition_graph_nodewise(
...
-            device=offsets.device,
-        )
-        global_src_ids_to_gpu = global_src_ids_per_rank // src_nodes_in_partition
-        remote_src_ids_per_rank = (
-            global_src_ids_per_rank - global_src_ids_to_gpu * src_nodes_in_partition
+            global_src_ids_on_rank.size(0),
+            dtype=local_offsets.dtype,
+            device=local_offsets.device,
         )
+        remote_src_ids_on_rank = mapping_global_src_ids_to_local_ids[
```
Comment (akshaysubr) [inline modulus/models/gnn_layers/distributed_graph.py:211]: Oh, I see the distinction between local and remote src nodes. Might be good to clarify this in the docstring

File: modulus/models/gnn_layers/distributed_graph.py [hunk 2]
```diff
@@ -223,6 +254,341 @@ def partition_graph_nodewise(
...
+        global_indices,
+        mapping_src_ids_to_ranks,
+        mapping_dst_ids_to_ranks,
+        partition_size,
+        partition_rank,
+        device,
+    )
+
+
+def partition_graph_coordinatewise(
```
Comment (akshaysubr) [inline modulus/models/gnn_layers/distributed_graph.py:334]: Would prefer if this was named slightly differently to make it explicit that you're specifying coordinate ranges. Maybe something like `partition_graph_by_coordinate_bounding_box`?

File: modulus/models/gnn_layers/distributed_graph.py [hunk 3]
```diff
@@ -223,6 +254,341 @@ def partition_graph_nodewise(
...
+    >>> num_dst_nodes = 4
+    >>> offsets = torch.arange(num_dst_nodes + 1, dtype=torch.int64) * 2
+    >>> indices = torch.arange(num_src_nodes, dtype=torch.int64)
+    >>>
+    >>> # example with 2D coordinates
+    >>> # assuming partitioning a 2D problem into the 4 quadrants
+    >>> partition_size = 4
+    >>> partition_rank = 0
+    >>> coordinate_separators_min = [[0, 0], [None, 0], [None, None], [0, None]]
+    >>> coordinate_separators_max = [[None, None], [0, None], [0, 0], [None, 0]]
```
Comment (akshaysubr) [inline modulus/models/gnn_layers/distributed_graph.py:387]: How is passing `None` allowed here when the type annotation is `List[List[float]]`?

File: modulus/models/gnn_layers/distributed_graph.py [hunk 4]
```diff
@@ -223,6 +254,341 @@ def partition_graph_nodewise(
...
+    >>>     indices,
+    >>>     src_coordinates,
+    >>>     dst_coordinates,
+    >>>     coordinate_separators_min,
+    >>>     coordinate_separators_max,
+    >>>     partition_size,
+    >>>     partition_rank,
+    >>>     device,
+    >>> )
+    GraphPartition(
```
Comment (akshaysubr) [inline modulus/models/gnn_layers/distributed_graph.py:482]: These examples are really nice! :)

File: test/models/meshgraphnet/test_meshgraphnet_snmg.py [hunk 5]
```diff
@@ -183,9 +182,10 @@ def run_test_distributed_meshgraphnet(rank, world_size, dtype):
     DistributedManager.cleanup()
 
 
+@import_or_fail("dgl")
 @pytest.mark.multigpu
-@pytest.mark.parametrize("dtype", [torch.float32, torch.float16, torch.bfloat16])
-def test_distributed_meshgraphnet(dtype):
+@pytest.mark.parametrize("dtype", [torch.float32, torch.float16])
+def test_distributed_meshgraphnet(dtype, pytestconfig):
```
Comment (akshaysubr) [inline test/models/meshgraphnet/test_meshgraphnet_snmg.py:188]: What is this `pytestconfig`? Doesn't seem to be used here?

File: test/models/graphcast/test_graphcast_snmg.py [hunk 6]
```diff
@@ -172,7 +172,7 @@ def run_test_distributed_graphcast(
 
 
 @pytest.mark.multigpu
-@pytest.mark.parametrize("dtype", [torch.float32, torch.float16, torch.bfloat16])
+@pytest.mark.parametrize("dtype", [torch.float32, torch.float16])
```
Comment (akshaysubr) [inline test/models/graphcast/test_graphcast_snmg.py:175]: Are the new partitioning schemes not included in these tests?

**Review by akshaysubr — COMMENTED** [review]
This PR is very nice! My main comment is that the new partitioning schemes are not tested at the moment. Other than that, only had some minor comments.

File: modulus/models/gnn_layers/distributed_graph.py [hunk 7]
```diff
@@ -71,33 +72,46 @@ def __post_init__(self):
...
+) -> GraphPartition:  # pragma: no cover
+    """
+    Utility function which partitions a global graph given as CSC structure.
+    It partitions both the global ID spaces for source nodes and destination nodes
+    based on the corresponding mappings as well as the graph structure and edge IDs.
+    Each rank maintains both a partition of the global source and destination nodes.
+    In terms of graph structure, each rank manages its own local graph structure
+    based on its partition of destination node IDs and all edges which - from the
+    point of view of each destination node on a current rank - are incoming edges.
+    For GNN operations this means, that features from source nodes need to be exchanged
```
[the remark this answers is not in the document]
Comment (stadlmax) [reply modulus/models/gnn_layers/distributed_graph.py:92]: I think this should be the same. All source node features are exchanged per se, but I think that NCCL should be smart enough to only exchange things which are not already available in a partition.

File: modulus/models/gnn_layers/distributed_graph.py (hunk 2 above)
Comment (stadlmax) [reply modulus/models/gnn_layers/distributed_graph.py:334]: good suggestion 👍🏽

**Review by akshaysubr — APPROVED** [review]

**Review by mnabian — APPROVED** [review]

**Pull request merged**